dev/PagerControl/PagerControl.cpp

```diff
@@ -0,0 +1,773 @@
+    {
+        winrt::VisualStateManager::GoToState(*this, c_comboBoxVisibleVisualState, false);
+    }
+    else if (displayMode == winrt::PagerControlDisplayMode::NumberBox)
+    {
+        winrt::VisualStateManager::GoToState(*this, c_numberBoxVisibleVisualState, false);
+    }
+    else
+    {
+        UpdateDisplayModeAutoState();
+    }
+}
+
+void PagerControl::UpdateDisplayModeAutoState()
+{
+    const int numberOfPages = NumberOfPages();
+    if (numberOfPages > -1)
+    {
+        winrt::VisualStateManager::GoToState(*this, numberOfPages < c_AutoDisplayModeNumberOfPagesThreshold ?
+            c_comboBoxVisibleVisualState : c_numberBoxVisibleVisualState, false);
+    }
+    else
+    {
+        winrt::VisualStateManager::GoToState(*this, c_numberBoxVisibleVisualState, false);
+    }
+}
+
+void PagerControl::OnNumberOfPagesChanged(const int oldValue)
+{
+    m_lastNumberOfPagesCount = oldValue;
+    const int numberOfPages = NumberOfPages();
+    if (numberOfPages < SelectedPageIndex() && numberOfPages > -1)
+    {
+        SelectedPageIndex(numberOfPages - 1);
+    }
+    UpdateTemplateSettingElementLists();
+    if (DisplayMode() == winrt::PagerControlDisplayMode::Auto)
+    {
+        UpdateDisplayModeAutoState();
+    }
+    if (numberOfPages > -1)
+    {
+        winrt::VisualStateManager::GoToState(*this, c_finiteItemsModeState, false);
+        if (const auto numberBox = m_numberBox.get())
+        {
+            numberBox.Maximum(numberOfPages);
+        }
+    }
+    else
+    {
+        winrt::VisualStateManager::GoToState(*this, c_infiniteItemsModeState, false);
+        if (const auto numberBox = m_numberBox.get())
+        {
+            numberBox.Maximum(INFINITY);
+        }
+    }
+    UpdateOnEdgeButtonVisualStates();
+}
+
+void PagerControl::OnSelectedPageIndexChange(const int oldValue)
+{
+    // If we don't have any pages, there is nothing we should do.
+    // Ensure that SelectedPageIndex will end up in the valid range of values
+    if (SelectedPageIndex() > NumberOfPages() - 1 && NumberOfPages() > -1)
+    {
+        SelectedPageIndex(NumberOfPages() - 1);
+    }
+    else if (SelectedPageIndex() < 0)
+    {
+        SelectedPageIndex(0);
+    }
+    // Now handle the value changes
+    m_lastSelectedPageIndex = oldValue;
+
+    if (const auto comboBox = m_comboBox.get())
+    {
+        if (SelectedPageIndex() < (int32_t)m_comboBoxEntries.Size())
+        {
+            comboBox.SelectedIndex(SelectedPageIndex());
+        }
+    }
+    if (const auto numBox = m_numberBox.get())
+    {
+        numBox.Value(SelectedPageIndex() + 1);
+    }
+
+    UpdateOnEdgeButtonVisualStates();
+    UpdateTemplateSettingElementLists();
+
+    if (DisplayMode() == winrt::PagerControlDisplayMode::ButtonPanel)
+    {
+        // NumberPanel needs to update based on multiple parameters.
+        // SelectedPageIndex is one of them, so let's do that now!
+        UpdateNumberPanel(NumberOfPages());
```

Comment:
If we switch to ButtonPanel mode we also update the number panel right?

Reply:
Yes, if we switch displaymode, we also update the necessary TemplateSettings collection.